File: web/src/soma/tokens.ts
// Anatomical tokens: fixed-layout numeric encoding of an assembled person.
// Port of soma/tokens.py.

import * as C from './constants';
import type { AnatomicalToken, Box, Person } from './types';

// Slot layout: sided parents get L/R slots, unsided parents one slot, +head.
export const SLOT_LAYOUT: Array<[number, number]> = (() => {
  const layout: Array<[number, number]> = [];
  for (const cid of C.KEYPOINT_PARENT_CLASS_IDS) {
    if (C.SIDE_PARENT_TO_CHILDREN.has(cid)) {
      layout.push([cid, 0]);
      layout.push([cid, 1]);
    } else {
      layout.push([cid, -1]);
    }
  }
  return layout;
})();
export const HEAD_SLOT = SLOT_LAYOUT.length;
export const N_SLOTS = SLOT_LAYOUT.length + 1; // keypoint slots + head

const SLOT_INDEX = new Map<string, number>(SLOT_LAYOUT.map(([cid, side], i) => [`${cid}:${side}`, i]));
const ANKLE_SLOTS = SLOT_LAYOUT.map(([cid], i) => (cid === 42 ? i : -1)).filter((i) => i >= 0);

function groupExtent(person: Person): Box | null {
  const xs: number[] = [];
  const ys: number[] = [];
  for (const kp of person.keypoints) {
    xs.push(kp.x);
    ys.push(kp.y);
  }
  if (person.headBox !== null) {
    xs.push(person.headBox[0], person.headBox[2]);
    ys.push(person.headBox[1], person.headBox[3]);
  }
  if (xs.length === 0) {
    return null;
  }
  return [Math.min(...xs), Math.min(...ys), Math.max(...xs), Math.max(...ys)];
}

export function buildToken(person: Person): AnatomicalToken | null {
  const boxProxy = person.bodyBox !== null ? person.bodyBox : groupExtent(person);
  if (boxProxy === null) {
    return null;
  }
  const w = Math.max(boxProxy[2] - boxProxy[0], 1.0);
  const h = Math.max(boxProxy[3] - boxProxy[1], 1.0);

  const presence = new Array<boolean>(N_SLOTS).fill(false);
  const points = new Float32Array(N_SLOTS * 2).fill(Number.NaN);

  // best-scoring keypoint per slot; unknown-side keypoints take a free side slot
  const sorted = [...person.keypoints].sort((a, b) => b.score - a.score);
  for (const kp of sorted) {
    const cid = kp.classId;
    let cand: Array<[number, number]>;
    if (C.SIDE_PARENT_TO_CHILDREN.has(cid)) {
      cand = kp.side === 0 || kp.side === 1 ? [[cid, kp.side]] : [[cid, 0], [cid, 1]];
    } else {
      cand = [[cid, -1]];
    }
    for (const [c, s] of cand) {
      const si = SLOT_INDEX.get(`${c}:${s}`);
      if (si !== undefined && !presence[si]) {
        presence[si] = true;
        points[si * 2] = (kp.x - boxProxy[0]) / w;
        points[si * 2 + 1] = (kp.y - boxProxy[1]) / h;
        break;
      }
    }
  }
  if (person.headBox !== null) {
    const hc = person.headBox;
    presence[HEAD_SLOT] = true;
    points[HEAD_SLOT * 2] = ((hc[0] + hc[2]) / 2 - boxProxy[0]) / w;
    points[HEAD_SLOT * 2 + 1] = ((hc[1] + hc[3]) / 2 - boxProxy[1]) / h;
  }

  return {
    anchor: person.anchor,
    bodyBox: person.bodyBox === null ? null : ([...person.bodyBox] as Box),
    bodyScore: person.bodyScore,
    boxProxy: [...boxProxy] as Box,
    presence,
    points,
    generation: person.generation,
    gender: person.gender,
    embedding: null, // SOMA-R: injected by the perception layer
    headBox: person.headBox === null ? null : ([...person.headBox] as Box),
    headDir: person.headDir === null ? null : [...person.headDir],
    headDirConf: person.headDirConf,
    crowding: 0,
  };
}

// Amodal full-body synthesis (v3): reconstruct the standing extent from
// anatomy (full height ~= alpha * head height); only ever EXTENDS the box
// downward. Port of soma/tokens.py apply_amodal.
export function applyAmodal(tok: AnatomicalToken, alpha: number, gamma = 0.9, overshoot = 1.0): void {
  if (tok.bodyBox === null || tok.headBox === null) {
    return;
  }
  const headH = tok.headBox[3] - tok.headBox[1];
  if (headH <= 2.0) {
    return;
  }
  if (ANKLE_SLOTS.some((s) => tok.presence[s])) {
    return; // visible ankles: the bottom edge is real
  }
  const y1 = Math.min(tok.bodyBox[1], tok.headBox[1]);
  const hEst = alpha * headH;
  if (tok.bodyBox[3] - y1 >= gamma * hEst) {
    return; // tall enough: no truncation evidence
  }
  const newY2 = y1 + overshoot * hEst;
  tok.bodyBox = [...tok.bodyBox] as Box;
  tok.bodyBox[1] = y1;
  tok.bodyBox[3] = Math.max(tok.bodyBox[3], newY2);
  tok.boxProxy = tok.bodyBox;
}

File: soma/tokens.py
"""Anatomical tokens: fixed-layout numeric encoding of an assembled person.

The token is the interchange format between the perception layer (L0) and
every association layer above it (SOMA-Zero cost channels, SOMA-Net input).
"""
from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from . import constants as C
from .assembly import Person
from .detector import Detections

# Slot layout: sided parents get L/R slots, unsided parents one slot, +head.
SLOT_LAYOUT: list[tuple[int, int]] = []   # (class_id, side) ; side -1 = unsided
for _cid in C.KEYPOINT_PARENT_CLASS_IDS:
    if _cid in C.SIDE_PARENT_TO_CHILDREN:
        SLOT_LAYOUT.append((_cid, 0))
        SLOT_LAYOUT.append((_cid, 1))
    else:
        SLOT_LAYOUT.append((_cid, -1))
HEAD_SLOT = len(SLOT_LAYOUT)
N_SLOTS = len(SLOT_LAYOUT) + 1           # keypoint slots + head

_SLOT_INDEX = {key: i for i, key in enumerate(SLOT_LAYOUT)}



@dataclass
class AnatomicalToken:
    anchor: str                     # body | wheelchair | crutches | orphan
    body_box: np.ndarray | None     # (4,) frame coords, None for orphan groups
    body_score: float
    box_proxy: np.ndarray           # (4,) always present: body box or part-group extent
    presence: np.ndarray            # (N_SLOTS,) bool
    points: np.ndarray              # (N_SLOTS, 2) normalized to box_proxy, NaN if absent
    generation: int
    gender: int
    embedding: np.ndarray | None = None   # SOMA-R: external ReID embedding, L2-normalized
    head_box: np.ndarray | None = None    # (4,) frame coords (amodal synthesis input)
    head_dir: np.ndarray | None = None    # (2,) unit (cos, sin) soft head orientation
    head_dir_conf: float = 0.0            # circular resultant length in [0,1]
    crowding: float = 0.0                 # max IoU with another body this frame


def _group_extent(person: Person) -> np.ndarray | None:
    xs = [kp.x for kp in person.keypoints]
    ys = [kp.y for kp in person.keypoints]
    if person.head_box is not None:
        xs += [person.head_box[0], person.head_box[2]]
        ys += [person.head_box[1], person.head_box[3]]
    if not xs:
        return None
    return np.array([min(xs), min(ys), max(xs), max(ys)], dtype=np.float32)


def build_token(person: Person, det: Detections) -> AnatomicalToken | None:
    box_proxy = person.body_box if person.body_box is not None else _group_extent(person)
    if box_proxy is None:
        return None
    w = max(float(box_proxy[2] - box_proxy[0]), 1.0)
    h = max(float(box_proxy[3] - box_proxy[1]), 1.0)

    presence = np.zeros(N_SLOTS, dtype=bool)
    points = np.full((N_SLOTS, 2), np.nan, dtype=np.float32)

    # best-scoring keypoint per slot; unknown-side keypoints take a free side slot
    for kp in sorted(person.keypoints, key=lambda k: -k.score):
        cid = kp.class_id
        if cid in C.SIDE_PARENT_TO_CHILDREN:
            cand = [(cid, kp.side)] if kp.side in (0, 1) else [(cid, 0), (cid, 1)]
        else:
            cand = [(cid, -1)]
        for key in cand:
            si = _SLOT_INDEX[key]
            if not presence[si]:
                presence[si] = True
                points[si] = ((kp.x - box_proxy[0]) / w, (kp.y - box_proxy[1]) / h)
                break
    if person.head_box is not None:
        hc = person.head_box
        presence[HEAD_SLOT] = True
        points[HEAD_SLOT] = (((hc[0] + hc[2]) / 2 - box_proxy[0]) / w,
                             ((hc[1] + hc[3]) / 2 - box_proxy[1]) / h)

    return AnatomicalToken(
        anchor=person.anchor,
        body_box=None if person.body_box is None else person.body_box.astype(np.float32),
        body_score=person.body_score,
        box_proxy=box_proxy.astype(np.float32),
        presence=presence,
        points=points,
        generation=person.generation,
        gender=person.gender,
        embedding=None,          # SOMA-R: injected by the perception layer
        head_box=None if person.head_box is None else person.head_box.astype(np.float32),
        head_dir=person.head_dir,
        head_dir_conf=person.head_dir_conf,
    )




# Amodal full-body synthesis (v3): MOT GT is amodal — occluded people keep
# their full-extent box. Our detector boxes hug the VISIBLE extent (exp007:
# det_h/gt_h p25 = 0.58 for vis<0.6), so we reconstruct the standing extent
# from anatomy: full height ~= alpha * head height. Applied at tracking time
# (sweepable without cache rebuilds); only ever EXTENDS the box downward.
_ANKLE_SLOTS = tuple(i for i, (cid, _s) in enumerate(SLOT_LAYOUT) if cid == 42)


def apply_amodal(tok: AnatomicalToken, alpha: float, gamma: float = 0.90,
                 overshoot: float = 1.0) -> None:
    if tok.body_box is None or tok.head_box is None:
        return
    head_h = float(tok.head_box[3] - tok.head_box[1])
    if head_h <= 2.0:
        return
    if any(tok.presence[s] for s in _ANKLE_SLOTS):
        return                       # visible ankles: the bottom edge is real
    y1 = min(float(tok.body_box[1]), float(tok.head_box[1]))
    h_est = alpha * head_h
    if (float(tok.body_box[3]) - y1) >= gamma * h_est:
        return                       # tall enough: no truncation evidence
    new_y2 = y1 + overshoot * h_est
    tok.body_box = tok.body_box.copy()
    tok.body_box[1] = y1
    tok.body_box[3] = max(float(tok.body_box[3]), new_y2)
    tok.box_proxy = tok.body_box

